python/examples/math/program.py

```diff
@@ -0,0 +1,166 @@
+    "@steps": FunctionCall[];
+}
+
+// A function call specifies a function name and a list of argument expressions. Arguments may contain
+// nested function calls and result references.
+export type FunctionCall = {
+    // Name of the function
+    "@func": string;
+    // Arguments for the function, if any
+    "@args"?: Expression[];
+};
+
+// An expression is a JSON value, a function call, or a reference to the result of a preceding expression.
+export type Expression = JsonValue | FunctionCall | ResultReference;
+
+// A JSON value is a string, a number, a boolean, null, an object, or an array. Function calls and result
+// references can be nested in objects and arrays.
+export type JsonValue = string | number | boolean | null | { [x: string]: Expression } | Expression[];
+
+// A result reference represents the value of an expression from a preceding step.
+export type ResultReference = {
+    // Index of the previous expression in the "@steps" array
+    "@ref": number;
+};
+'''
+
+
+ResultReference = TypedDict(
+    "ResultReference", {"@ref": Annotated[int, Doc("Index of the previous expression in the 'steps' array")]}
+)
+
+FunctionCall = TypedDict(
+    "FunctionCall",
+    {
+        "@func": Annotated[str, Doc("Name of the function")],
+        "@args": NotRequired[Annotated[list["Expression"], Doc("Arguments for the function, if any")]],
+    },
+)
+
+JsonValue = str | int | float | bool | None | dict[str, "Expression"] | list["Expression"]
+Expression = JsonValue | FunctionCall | ResultReference
+
+JsonProgram = TypedDict("Program", {"@steps": list[FunctionCall]})
+
+
+async def evaluate_json_program(program: Mapping[str, Any], onCall: Callable[[str, list[Any]], Awaitable[Any]]) -> Any:
+    results: list[Any] = []
+
+    async def evaluate_array(array: list[Any]) -> list[Any]:
+        return await asyncio.gather(*[evaluate_call(e) for e in array])
+
+    async def evaluate_object(expr: FunctionCall):
+        if "@ref" in expr:
+            index = expr["@ref"]
+            if index < len(results):
+                return results[index]
+
+        elif "@func" in expr and "@args" in expr:
+            function_name = expr["@func"]
+            return await onCall(function_name, await evaluate_array(expr["@args"]))
+
+        elif isinstance(expr, collections.abc.Sequence):
+            return await evaluate_array(expr)
+
+        else:
+            raise ValueError("This condition should never hit")
+
+    async def evaluate_call(expr: FunctionCall):
+        # implement awaitable pattern
+        if isinstance(expr, int) or isinstance(expr, float) or isinstance(expr, str):
+            return expr
+        return await evaluate_object(expr)
+
+    for step in program["@steps"]:
+        results.append(await evaluate_call(step))
+
+    if len(results) > 0:
+        return results[-1]
+    else:
+        return None
+
+
+class TypeChatProgramValidator(TypeChatValidator[T]):
+    def __init__(self, py_type: type[T]):
```

Comment:
**Type** instead of type

Reply:
Might be mistaken, but I do think it should be `type`